### ISSUE TITLE
AddGradientOperators for Regression

### ISSUE THREAD

**Issue by RyanMorris95** [facebookarchive/caffe2#385]
Hi Everyone,

I have been trying to build off of the MNIST tutorial but facial landmark detection.  So, now I want my network to regress to the pixel coordinates of the landmarks.  However, when I create the AddGradientOperators I get thrown this exception: 

`Exception: No gradient registered for Cast. Exception from creating the gradient op: [enforce fail at operator_gradient.h:279] false. Operator Cast should have a gradient but is not implemented yet. .`

I am pretty sure that just means that the gradient operator only accepts floats for right now.  So I tried casting my loss blob to a float and still got thrown the exception.  Here is my relevant code:
```
def AddTrainingOperators(model, pred, label):
    # compute the expected loss
    dist = model.SquaredL2Distance([label, pred], "dist")
    loss = dist.AveragedLoss([], ["loss"])
    loss = model.Cast(loss, "loss", to=core.DataType.FLOAT)
    AddAccuracy(model, pred, label)
    # use the average loss we just computer to add gradient operators
    model.AddGradientOperators([loss])
```

Here is the code from the toy regression problem:
```

# The loss function is computed by a squared L2 distance, and then averaged
# over all items in the minibatch.
dist = train_net.SquaredL2Distance([Y_noise, Y_pred], "dist")
loss = dist.AveragedLoss([], ["loss"])
# Get gradients for all the computations above.
gradient_map = train_net.AddGradientOperators([loss])

```

What am I doing wrong here?  Or is this a bug?

Thank you for your time!

**Comment by harouwu** [comment]
Hi @RyanMorris95 , great to hear you are getting your hands dirty with Caffe2. The type C2 operators take can be various. Some ops will take INT32 as input. I looked at the code and I found that maybe you want to do

```
loss = model.AveragedLoss(dist, "loss")
```

and get rid of the cast. Because the output of AveragedLoss is already in float type.

Hope this will solve your questions :)

**Comment by harouwu** [comment]
Adding an operator(layer if it is more familiar) to caffe2 model is like:

```
model.<OpName>([input_list], [output_list], *args, **kwargs)
```

We also provide some helper functions that will make model building easier. We will name it `brew` and it is coming out very soon. And we will have tutorial for that too. Please keep in touch

**Comment by RyanMorris95** [comment]
@harouwu Thank you for the reply and the help!  Along with deleting my accuracy function that fixed the issue.  However I am getting a new error always at Iteration number 853.  Here is the error output:

```
*** Aborted at 1492972198 (unix time) try "date -d @1492972198" if you are using GNU date ***
PC: @     0x7fa19367f6ed caffe2::Tensor<>::ResizeLike<>()
*** SIGSEGV (@0x0) received by PID 14532 (TID 0x7fa1ba6a8700) from PID 0; stack trace: ***
    @     0x7fa1b9ec6390 (unknown)
    @     0x7fa19367f6ed caffe2::Tensor<>::ResizeLike<>()
    @     0x7fa19368d511 caffe2::CastOp<>::DoRunWithDstType<>()
    @     0x7fa19a204673 caffe2::Operator<>::Run()
    @     0x7fa193624448 caffe2::SimpleNet::Run()
    @     0x7fa1936193f0 caffe2::Workspace::RunNet()
    @     0x7fa19a1dfb07 _ZZN8pybind1112cpp_function10initializeIZN6caffe26python16addGlobalMethodsERNS_6moduleEEUlRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiE13_bJSD_iEJNS_4nameENS_7siblingENS_5scopeEEEEvOT_PFT0_DpT1_EDpRKT2_ENKUlPNS_6detail15function_recordENS_6handleESW_SW_E1_clESV_SW_SW_SW_.constprop.1550
    @     0x7fa19a1dfd47 _ZZN8pybind1112cpp_function10initializeIZN6caffe26python16addGlobalMethodsERNS_6moduleEEUlRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiE13_bJSD_iEJNS_4nameENS_7siblingENS_5scopeEEEEvOT_PFT0_DpT1_EDpRKT2_ENUlPNS_6detail15function_recordENS_6handleESW_SW_E1_4_FUNESV_SW_SW_SW_
    @     0x7fa19a205818 pybind11::cpp_function::dispatcher()
    @     0x7fa1ba1cf5a2 PyEval_EvalFrameEx
    @     0x7fa1ba1d01ce PyEval_EvalCodeEx
    @     0x7fa1ba1cf1f6 PyEval_EvalFrameEx
    @     0x7fa1ba1d01ce PyEval_EvalCodeEx
    @     0x7fa1ba1d02e2 PyEval_EvalCode
    @     0x7fa1ba1f0960 PyRun_FileExFlags
    @     0x7fa1ba1f0b3f PyRun_SimpleFileExFlags
    @     0x7fa1ba206484 Py_Main
    @     0x7fa1b93fc830 __libc_start_main
    @           0x400649 (unknown)
```
I do not know what this error is.  Maybe I have some label in the wrong format?

Appreciate the help!

**Comment by RyanMorris95** [comment]
The example you provided of the operator format has helped me with my understanding, so thanks!  Also, I am excited to see what `brew` has in store for us!

**Comment by harouwu** [comment]
This problem seems related to Cast Operator. Would you mind show me where Cast is used in your code?

**Comment by RyanMorris95** [comment]
Here is where I use Cast:

```
def AddInput(model, batch_size, db, db_type):
    """
    Creates input graph.
    :param model: Neural network model
    :param batch_size: Number of images to run at once.
    :param db: Database of images and labels
    :param db_type: Type of database
    :return: data and label
    """
    data_uint8, label = model.TensorProtosDBInput(
        [], ["data_uint8", "label"], batch_size=batch_size,
        db=db, db_type=db_type
    )
    # Cast the data to float
    data = model.Cast(data_uint8, "data", to=core.DataType.FLOAT)
    # Don't need the gradient for the backward pass
    data = model.StopGradient(data, data)
    return data, label

```

**Comment by harouwu** [comment]
From my understanding, this is related to a invalid input. If your situation is that this problem won't happen until certain amount of data is consumed. I guess maybe you want to check:

- If the input data is in a database form. And make sure you passed the right db type.
- If your data is already in float, then remove the CastOp at all.

I think we might need to have more tutorial about reading data.
